### PR TITLE
Add fast copy-constructor for adding index to Metadata

### PR DESCRIPTION
**Pull request by original-brownbear**
Add a fast copy constructor for creating an index. The building of new metadata is taking more than 15% of CPU time on the master node during many shards benchmark bootstrapping and completely disappears from profiling with this change.

It's unfortunate that we don't really have a nice way of testing this logic directly since we don't have a clean equality check on metadata but it has massive coverage through its use for almost all index creates (except for those using a builder filter function) in all ITs so this seems safe enough to me.

### CODE REVIEW

**Comment by DaveCTurner** [comment]
Is this still something we're pursuing @original-brownbear?

**Comment by original-brownbear** [comment]
Yea let me clean this up real quick, with all the recent fixes this one is a 10%+ speedup in benchmark bootstrap time actually ... would be nice to have this

**Review by DaveCTurner — APPROVED** [review]
LGTM, a couple of nits

It'd be nice to have a way to assert that a `Metadata` satisfies all the invariants we expect of it just to make sure we don't mess anything up when adding copy-constructors like this (now or in future)

File: server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java [hunk 1]
```diff
@@ -439,6 +440,136 @@ public Metadata withAllocationAndTermUpdatesOnly(Map<String, IndexMetadata> upda
...
+            updatedOpenIndices = ArrayUtils.concat(allOpenIndices, indexName);
+            if (index.isHidden() == false) {
+                updatedVisibleOpenIndices = ArrayUtils.concat(visibleOpenIndices, indexName);
+            } else {
+                updatedVisibleOpenIndices = visibleOpenIndices;
+            }
+            updatedVisibleClosedIndices = visibleClosedIndices;
+            updatedClosedIndices = allClosedIndices;
+        } else {
+            assert index.getState() == IndexMetadata.State.CLOSE
```
[the remark this answers is not in the document]
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java]: Both options look weird, but I think I like a switch better as well => moved to switch :)

**Comment by original-brownbear** [comment]
Thanks David, I applied your NITs and I agree on the assertions. I'll follow up for this, I'd like to do a bigger PR here where we just move all the asserts from the current builder to the constructor, that should give us a neat safety net.

**Pull request merged**